Add `Keymap`

KDecay opened the pull request:
I added a basic `Keymap` implementation with two examples. This should definitely be expanded to also be able to read keybindings from files using a `from_file` method, but I didn't want to add everything all at once.

Currently the `Keymap` supports inserting and removing keybindings, both during runtime and during the creation of the model.

rhelmot reviewed (changes requested):
My main concern here is that we're trying to cover every GUI application which could ever be built with vizia, so we want the most general possible interface here, and to me that means making keystroke <--> action a many-to-many relationship. I would also prefer if the actions didn't have to be eq + hash, so maybe the core storage should maybe be `HashMap<Key, Vec<Action>>`? For the case where you're trying to build a keymap editing view, there can be an `export` function that transforms the map into a `Vec<(Action, Vec<Key>)>`.

KDecay commented on core/src/input/binding.rs:
```diff
@@ -0,0 +1,25 @@
+
+/// A keybinding used inside of a [`Keymap`](crate::Keymap).
+#[derive(Default, Copy, Clone, Debug)]
+pub struct KeyBinding {
```
Very true. I'll change this as well.

KDecay commented:
I renamed the `KeyBinding` to `KeyChord` and also updated the `Keymap` to allow for many-to-many relationships. It is now able to just lookup a `KeyChord` and retrieve a `Vec` over all actions that get triggered through that `KeyChord`. So instead of having to lookup every action we now just get the actions that are being pressed.

rhelmot commented on core/src/input/keymap.rs:
```diff
@@ -105,131 +106,174 @@ where
-    pub fn remove(mut self, key: &T) -> Self {
-        self.bindings.remove(key);
-        self
+    pub fn pressed_actions(&self, cx: &Context, code: Code) -> Option<impl Iterator<Item = &T>> {
```
Why is this an Option<impl Iterator> and not just an iterator which is sometimes empty?

KDecay replied:
That's because I'm returning an iterator over `&T` which in turn doesn't allow me to return something like `Vec::new().iter()`, because that value would be dropped at the end of the scope. Is there a way to get around this that you know of?

rhelmot replied:
Can you do something like `actions.get(...).iter().flat_map(|actions| actions.iter())`? Option has an iter method which iterates over the single item if it's Some, and nothing if it's None.

KDecay replied:
I couldn't make your suggestion work, because it would return a `FlatMap` but I found another solution is just to use [].iter() in the else case ^^ Now it just returns an `impl Iterator<Item = &T>` without an `Option` around it.
